File: server/src/kfchess/game/moves.py
"""Move definitions and validation for Real-time-chess-battle (中国象棋版)."""

from __future__ import annotations

from dataclasses import dataclass, field

from kfchess.game.board import Board
from kfchess.game.pieces import Piece, PieceType

PathPoint = tuple[float, float]


@dataclass
class _EnemyMoveInfo:
    dr: int
    dc: int
    start_r: int
    start_c: int
    forward_squares: list[tuple[int, int]]


@dataclass
class PathClearContext:
    own_forward_path: set[tuple[int, int]] = field(default_factory=set)
    moving_piece_ids: set[str] = field(default_factory=set)
    enemy_moves: list[_EnemyMoveInfo] = field(default_factory=list)


FOUR_PLAYER_ORIENTATIONS: dict[int, object] = {}


@dataclass
class Move:
    piece_id: str
    path: list[PathPoint]
    start_tick: int
    extra_move: Move | None = None

    def to_dict(self) -> dict:
        return {
            "piece_id": self.piece_id,
            "path": [list(p) for p in self.path],
            "start_tick": self.start_tick,
            "extra_move": self.extra_move.to_dict() if self.extra_move else None,
        }

    @classmethod
    def from_dict(cls, data: dict) -> Move:
        return cls(
            piece_id=data["piece_id"],
            path=[tuple(p) for p in data["path"]],
            start_tick=data["start_tick"],
            extra_move=cls.from_dict(data["extra_move"]) if data.get("extra_move") else None,
        )

    @property
    def start_position(self) -> PathPoint:
        return self.path[0]

    @property
    def end_position(self) -> PathPoint:
        return self.path[-1]

    @property
    def num_squares(self) -> int:
        return len(self.path) - 1


@dataclass
class Cooldown:
    piece_id: str
    start_tick: int
    duration: int

    def to_dict(self) -> dict:
        return {
            "piece_id": self.piece_id,
            "start_tick": self.start_tick,
            "duration": self.duration,
        }

    @classmethod
    def from_dict(cls, data: dict) -> Cooldown:
        return cls(
            piece_id=data["piece_id"],
            start_tick=data["start_tick"],
            duration=data["duration"],
        )

    def is_active(self, current_tick: int) -> bool:
        return current_tick < self.start_tick + self.duration


def build_path_clear_context(
    player: int,
    board: Board,
    active_moves: list[Move],
    current_tick: int,
    ticks_per_square: int,
) -> PathClearContext:
    del player, board, current_tick, ticks_per_square
    return PathClearContext(moving_piece_ids={m.piece_id for m in active_moves})


def compute_move_path(
    piece: Piece,
    board: Board,
    to_row: int,
    to_col: int,
    active_moves: list[Move],
    current_tick: int = 0,
    ticks_per_square: int = 30,
    path_context: PathClearContext | None = None,
) -> list[PathPoint] | None:
    del current_tick, ticks_per_square
    from_row, from_col = piece.grid_position
    if (from_row, from_col) == (to_row, to_col):
        return None
    if not board.is_valid_square(to_row, to_col):
        return None

    moving_ids = path_context.moving_piece_ids if path_context else {m.piece_id for m in active_moves}
    destination_piece = board.get_piece_at(to_row, to_col)
    if destination_piece is not None and destination_piece.player == piece.player and destination_piece.id not in moving_ids:
        return None

    match piece.type:
        case PieceType.SOLDIER | PieceType.PAWN:
            path = _compute_soldier_path(piece, board, from_row, from_col, to_row, to_col)
        case PieceType.HORSE | PieceType.KNIGHT:
            path = _compute_horse_path(piece, board, from_row, from_col, to_row, to_col, moving_ids)
        case PieceType.ELEPHANT | PieceType.BISHOP:
            path = _compute_elephant_path(piece, board, from_row, from_col, to_row, to_col, moving_ids)
        case PieceType.CHARIOT | PieceType.ROOK:
            path = _compute_chariot_path(piece, board, from_row, from_col, to_row, to_col, moving_ids)
        case PieceType.CANNON:
            path = _compute_cannon_path(piece, board, from_row, from_col, to_row, to_col, moving_ids)
        case PieceType.ADVISOR | PieceType.QUEEN:
            path = _compute_advisor_path(piece, board, from_row, from_col, to_row, to_col)
        case PieceType.GENERAL | PieceType.KING:
            path = _compute_general_path(piece, board, from_row, from_col, to_row, to_col, moving_ids)
        case _:
            path = None

    if path is None:
        return None
    if _would_generals_face_after_move(piece, board, to_row, to_col, moving_ids):
        return None
    return path


def should_promote_pawn(piece: Piece, board: Board, end_row: int, end_col: int) -> bool:
    del piece, board, end_row, end_col
    return False


def check_castling(
    piece: Piece,
    board: Board,
    to_row: int,
    to_col: int,
    active_moves: list[Move],
    cooldowns: list[Cooldown] | None = None,
    current_tick: int = 0,
    ticks_per_square: int = 30,
) -> tuple[Move, Move] | None:
    del piece, board, to_row, to_col, active_moves, cooldowns, current_tick, ticks_per_square
    return None


def _get_occupant(board: Board, row: int, col: int, moving_ids: set[str]) -> Piece | None:
    piece = board.get_piece_at(row, col)
    if piece is None:
        return None
    if piece.id in moving_ids:
        return None
    return piece


def _line_squares(from_row: int, from_col: int, to_row: int, to_col: int) -> list[tuple[int, int]]:
    squares: list[tuple[int, int]] = []
    if from_row == to_row:
        step = 1 if to_col > from_col else -1
        for c in range(from_col + step, to_col, step):
            squares.append((from_row, c))
    elif from_col == to_col:
        step = 1 if to_row > from_row else -1
        for r in range(from_row + step, to_row, step):
            squares.append((r, from_col))
    return squares


def _would_generals_face_after_move(piece: Piece, board: Board, to_row: int, to_col: int, moving_ids: set[str]) -> bool:
    generals: dict[int, tuple[int, int]] = {}
    for p in board.get_active_pieces():
        if p.captured:
            continue
        if p.type not in (PieceType.GENERAL, PieceType.KING):
            continue
        if p.id == piece.id:
            generals[p.player] = (to_row, to_col)
            continue
        if p.grid_position == (to_row, to_col) and p.player != piece.player:
            continue
        generals[p.player] = p.grid_position

    if 1 not in generals or 2 not in generals:
        return False
    red = generals[1]
    black = generals[2]
    if red[1] != black[1]:
        return False
    col = red[1]
    start = min(red[0], black[0]) + 1
    end = max(red[0], black[0])
    for row in range(start, end):
        occ = _get_occupant(board, row, col, moving_ids)
        if occ is None:
            continue
        if occ.id == piece.id:
            continue
        if occ.grid_position == (to_row, to_col) and occ.player != piece.player:
            continue
        return False
    return True


def _compute_soldier_path(piece: Piece, board: Board, from_row: int, from_col: int, to_row: int, to_col: int) -> list[PathPoint] | None:
    dr = to_row - from_row
    dc = to_col - from_col
    forward = -1 if piece.player == 1 else 1
    crossed = board.has_crossed_river(piece.player, from_row)
    allowed = {(forward, 0)}
    if crossed:
        allowed |= {(0, -1), (0, 1)}
    if (dr, dc) not in allowed:
        return None
    return [(from_row, from_col), (to_row, to_col)]


def _compute_horse_path(piece: Piece, board: Board, from_row: int, from_col: int, to_row: int, to_col: int, moving_ids: set[str]) -> list[PathPoint] | None:
    del piece
    dr = to_row - from_row
    dc = to_col - from_col
    if sorted((abs(dr), abs(dc))) != [1, 2]:
        return None
    if abs(dr) == 2:
        leg = (from_row + dr // 2, from_col)
    else:
        leg = (from_row, from_col + dc // 2)
    if _get_occupant(board, leg[0], leg[1], moving_ids) is not None:
        return None
    return [(from_row, from_col), leg, (to_row, to_col)]


def _compute_elephant_path(piece: Piece, board: Board, from_row: int, from_col: int, to_row: int, to_col: int, moving_ids: set[str]) -> list[PathPoint] | None:
    dr = to_row - from_row
    dc = to_col - from_col
    if abs(dr) != 2 or abs(dc) != 2:
        return None
    if not board.same_side_of_river(piece.player, to_row):
        return None
    eye = (from_row + dr // 2, from_col + dc // 2)
    if _get_occupant(board, eye[0], eye[1], moving_ids) is not None:
        return None
    return [(from_row, from_col), eye, (to_row, to_col)]


def _compute_advisor_path(piece: Piece, board: Board, from_row: int, from_col: int, to_row: int, to_col: int) -> list[PathPoint] | None:
    if abs(to_row - from_row) != 1 or abs(to_col - from_col) != 1:
        return None
    if not board.is_palace_square(piece.player, to_row, to_col):
        return None
    return [(from_row, from_col), (to_row, to_col)]


def _compute_general_path(piece: Piece, board: Board, from_row: int, from_col: int, to_row: int, to_col: int, moving_ids: set[str]) -> list[PathPoint] | None:
    destination_piece = board.get_piece_at(to_row, to_col)
    # 飞将：在同一列无阻挡时可以直接吃掉对方将/帅
    if destination_piece is not None and destination_piece.player != piece.player and destination_piece.type in (PieceType.GENERAL, PieceType.KING):
        if from_col != to_col:
            return None
        middle = _line_squares(from_row, from_col, to_row, to_col)
        blockers = [sq for sq in middle if _get_occupant(board, sq[0], sq[1], moving_ids) is not None]
        if blockers:
            return None
        return [(from_row, from_col), (to_row, to_col)]

    if abs(to_row - from_row) + abs(to_col - from_col) != 1:
        return None
    if not board.is_palace_square(piece.player, to_row, to_col):
        return None
    return [(from_row, from_col), (to_row, to_col)]


def _compute_chariot_path(piece: Piece, board: Board, from_row: int, from_col: int, to_row: int, to_col: int, moving_ids: set[str]) -> list[PathPoint] | None:
    del piece
    if from_row != to_row and from_col != to_col:
        return None
    middle = _line_squares(from_row, from_col, to_row, to_col)
    if any(_get_occupant(board, r, c, moving_ids) is not None for r, c in middle):
        return None
    return [(from_row, from_col), (to_row, to_col)]


def _compute_cannon_path(piece: Piece, board: Board, from_row: int, from_col: int, to_row: int, to_col: int, moving_ids: set[str]) -> list[PathPoint] | None:
    del piece
    if from_row != to_row and from_col != to_col:
        return None
    middle = _line_squares(from_row, from_col, to_row, to_col)
    blockers = sum(1 for r, c in middle if _get_occupant(board, r, c, moving_ids) is not None)
    destination_piece = _get_occupant(board, to_row, to_col, moving_ids)
    if destination_piece is None:
        if blockers != 0:
            return None
    else:
        if blockers != 1:
            return None
    return [(from_row, from_col), (to_row, to_col)]

File: client/src/game/moves.ts

/**
 * Client-side legal move calculation for Real-time-chess-battle（中国象棋）.
 */

import type { Piece, ActiveMove, BoardType } from '../stores/game';
import { BOARD_DIMENSIONS, isValidSquare } from './constants';

function getPieceAtLocation(pieces: Piece[], row: number, col: number): Piece | null {
  return pieces.find((p) => !p.captured && p.row === row && p.col === col) ?? null;
}

function isMoving(activeMoves: ActiveMove[], pieceId: string): boolean {
  return activeMoves.some((m) => m.pieceId === pieceId);
}

function isOccupiedByFriendly(pieces: Piece[], activeMoves: ActiveMove[], piece: Piece, row: number, col: number): boolean {
  const target = getPieceAtLocation(pieces, row, col);
  return !!target && target.player === piece.player && !isMoving(activeMoves, target.id);
}

function countLineBlockers(pieces: Piece[], activeMoves: ActiveMove[], fromRow: number, fromCol: number, toRow: number, toCol: number): number {
  if (fromRow !== toRow && fromCol !== toCol) return -1;
  let count = 0;
  if (fromRow === toRow) {
    const step = toCol > fromCol ? 1 : -1;
    for (let c = fromCol + step; c !== toCol; c += step) {
      const occ = getPieceAtLocation(pieces, fromRow, c);
      if (occ && !isMoving(activeMoves, occ.id)) count++;
    }
  } else {
    const step = toRow > fromRow ? 1 : -1;
    for (let r = fromRow + step; r !== toRow; r += step) {
      const occ = getPieceAtLocation(pieces, r, fromCol);
      if (occ && !isMoving(activeMoves, occ.id)) count++;
    }
  }
  return count;
}

function isPalaceSquare(player: number, row: number, col: number): boolean {
  if (col < 3 || col > 5) return false;
  return player === 1 ? row >= 7 && row <= 9 : row >= 0 && row <= 2;
}

function hasCrossedRiver(player: number, row: number): boolean {
  return player === 1 ? row <= 4 : row >= 5;
}

function generalsFacing(pieces: Piece[], activeMoves: ActiveMove[], movingPiece: Piece, toRow: number, toCol: number): boolean {
  let red: [number, number] | null = null;
  let black: [number, number] | null = null;
  for (const p of pieces) {
    if (p.captured) continue;
    let row = p.row;
    let col = p.col;
    if (p.id === movingPiece.id) {
      row = toRow;
      col = toCol;
    }
    if (p.type === 'G' && p.player === 1) red = [row, col];
    if (p.type === 'G' && p.player === 2) black = [row, col];
  }
  if (!red || !black) return false;
  if (red[1] !== black[1]) return false;
  for (let r = Math.min(red[0], black[0]) + 1; r < Math.max(red[0], black[0]); r++) {
    const occ = getPieceAtLocation(pieces, r, red[1]);
    if (!occ || occ.id === movingPiece.id || isMoving(activeMoves, occ.id)) continue;
    return false;
  }
  return true;
}

function isSoldierMove(piece: Piece, toRow: number, toCol: number): boolean {
  const dr = toRow - piece.row;
  const dc = toCol - piece.col;
  const forward = piece.player === 1 ? -1 : 1;
  if (dr === forward && dc === 0) return true;
  if (hasCrossedRiver(piece.player, piece.row) && dr === 0 && Math.abs(dc) === 1) return true;
  return false;
}

function isHorseMove(pieces: Piece[], activeMoves: ActiveMove[], piece: Piece, toRow: number, toCol: number): boolean {
  const dr = toRow - piece.row;
  const dc = toCol - piece.col;
  if (!((Math.abs(dr) === 2 && Math.abs(dc) === 1) || (Math.abs(dr) === 1 && Math.abs(dc) === 2))) return false;
  const legRow = Math.abs(dr) === 2 ? piece.row + dr / 2 : piece.row;
  const legCol = Math.abs(dc) === 2 ? piece.col + dc / 2 : piece.col;
  const leg = getPieceAtLocation(pieces, legRow, legCol);
  return !(leg && !isMoving(activeMoves, leg.id));
}

function isElephantMove(pieces: Piece[], activeMoves: ActiveMove[], piece: Piece, toRow: number, toCol: number): boolean {
  const dr = toRow - piece.row;
  const dc = toCol - piece.col;
  if (Math.abs(dr) !== 2 || Math.abs(dc) !== 2) return false;
  if (piece.player === 1 && toRow < 5) return false;
  if (piece.player === 2 && toRow > 4) return false;
  const eye = getPieceAtLocation(pieces, piece.row + dr / 2, piece.col + dc / 2);
  return !(eye && !isMoving(activeMoves, eye.id));
}

function isAdvisorMove(piece: Piece, toRow: number, toCol: number): boolean {
  return Math.abs(toRow - piece.row) === 1 && Math.abs(toCol - piece.col) === 1 && isPalaceSquare(piece.player, toRow, toCol);
}

function isGeneralMove(pieces: Piece[], activeMoves: ActiveMove[], piece: Piece, toRow: number, toCol: number): boolean {
  const target = getPieceAtLocation(pieces, toRow, toCol);
  if (target && target.player !== piece.player && target.type === 'G') {
    if (piece.col !== toCol) return false;
    return countLineBlockers(pieces, activeMoves, piece.row, piece.col, toRow, toCol) === 0;
  }
  return Math.abs(toRow - piece.row) + Math.abs(toCol - piece.col) === 1 && isPalaceSquare(piece.player, toRow, toCol);
}

function isChariotMove(pieces: Piece[], activeMoves: ActiveMove[], piece: Piece, toRow: number, toCol: number): boolean {
  return countLineBlockers(pieces, activeMoves, piece.row, piece.col, toRow, toCol) === 0;
}

function isCannonMove(pieces: Piece[], activeMoves: ActiveMove[], piece: Piece, toRow: number, toCol: number): boolean {
  const blockers = countLineBlockers(pieces, activeMoves, piece.row, piece.col, toRow, toCol);
  if (blockers < 0) return false;
  const target = getPieceAtLocation(pieces, toRow, toCol);
  if (!target || isMoving(activeMoves, target.id)) return blockers === 0;
  return target.player !== piece.player && blockers === 1;
}

export function isLegalMove(
  pieces: Piece[],
  activeMoves: ActiveMove[],
  _currentTick: number,
  _ticksPerSquare: number,
  piece: Piece,
  toRow: number,
  toCol: number,
  boardType: BoardType = 'standard'
): boolean {
  const dims = BOARD_DIMENSIONS[boardType];
  if (toRow < 0 || toRow >= dims.height || toCol < 0 || toCol >= dims.width) return false;
  if (!isValidSquare(toRow, toCol, boardType)) return false;
  if (piece.row === toRow && piece.col === toCol) return false;
  if (isOccupiedByFriendly(pieces, activeMoves, piece, toRow, toCol)) return false;

  let ok = false;
  switch (piece.type) {
    case 'P': ok = isSoldierMove(piece, toRow, toCol); break;
    case 'N': ok = isHorseMove(pieces, activeMoves, piece, toRow, toCol); break;
    case 'E': ok = isElephantMove(pieces, activeMoves, piece, toRow, toCol); break;
    case 'R': ok = isChariotMove(pieces, activeMoves, piece, toRow, toCol); break;
    case 'A': ok = isAdvisorMove(piece, toRow, toCol); break;
    case 'G': ok = isGeneralMove(pieces, activeMoves, piece, toRow, toCol); break;
    case 'C': ok = isCannonMove(pieces, activeMoves, piece, toRow, toCol); break;
    default: ok = false;
  }
  if (!ok) return false;
  return !generalsFacing(pieces, activeMoves, piece, toRow, toCol);
}

export function getLegalMovesForPiece(
  pieces: Piece[],
  activeMoves: ActiveMove[],
  currentTick: number,
  ticksPerSquare: number,
  piece: Piece,
  boardType: BoardType = 'standard'
): [number, number][] {
  const dims = BOARD_DIMENSIONS[boardType];
  const moves: [number, number][] = [];
  for (let row = 0; row < dims.height; row++) {
    for (let col = 0; col < dims.width; col++) {
      if (isLegalMove(pieces, activeMoves, currentTick, ticksPerSquare, piece, row, col, boardType)) {
        moves.push([row, col]);
      }
    }
  }
  return moves;
}

export function getAllLegalMoves(
  pieces: Piece[],
  activeMoves: ActiveMove[],
  currentTick: number,
  ticksPerSquare: number,
  playerNumber: number,
  boardType: BoardType = 'standard'
): Map<string, [number, number][]> {
  const allMoves = new Map<string, [number, number][]>();
  for (const piece of pieces) {
    if (piece.player === playerNumber && !piece.captured && !piece.moving && !piece.onCooldown) {
      const moves = getLegalMovesForPiece(pieces, activeMoves, currentTick, ticksPerSquare, piece, boardType);
      if (moves.length > 0) allMoves.set(piece.id, moves);
    }
  }
  return allMoves;
}
